nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/main/java/org/apache/nifi/processors/pulsar/pubsub/ConsumePulsarRecord.java

```diff
@@ -0,0 +1,324 @@
+        RELATIONSHIPS = Collections.unmodifiableSet(relationships);
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return RELATIONSHIPS;
+    }
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return PROPERTIES;
+    }
+
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
+        final RecordReaderFactory readerFactory = context.getProperty(RECORD_READER)
+                .asControllerService(RecordReaderFactory.class);
+
+        final RecordSetWriterFactory writerFactory = context.getProperty(RECORD_WRITER)
+                .asControllerService(RecordSetWriterFactory.class);
+
+        try {
+            Consumer<byte[]> consumer = getConsumer(context, getConsumerId(context, session.get()));
+
+            if (consumer == null) { /* If we aren't connected to Pulsar, then just yield */
+                context.yield();
+                return;
+            }
+
+            if (context.getProperty(ASYNC_ENABLED).isSet() && context.getProperty(ASYNC_ENABLED).asBoolean()) {
+               consumeAsync(consumer, context, session);
+               handleAsync(context, session, consumer, readerFactory, writerFactory);
+            } else {
+               consumeMessage(session, consumer, consumer.receive(), readerFactory, writerFactory);
+            }
+        } catch (PulsarClientException e) {
+            getLogger().error("Unable to consume from Pulsar Topic ", e);
+            context.yield();
+            throw new ProcessException(e);
+        }
+    }
+
+    private void consumeMessage(ProcessSession session, final Consumer<byte[]> consumer, final Message<byte[]> msg,
+            final RecordReaderFactory readerFactory, RecordSetWriterFactory writerFactory) throws PulsarClientException {
+        consumeMessage(session, consumer, msg, false, readerFactory, writerFactory);
+    }
+
+    /**
+     * Consumes a single message, parses it, and returns.
+     */
+    private void consumeMessage(ProcessSession session, final Consumer<byte[]> consumer, final Message<byte[]> msg, boolean async,
+         final RecordReaderFactory readerFactory, RecordSetWriterFactory writerFactory) throws PulsarClientException {
+
+       RecordSetWriter writer = null;
+       FlowFile flowFile = session.create();
+       OutputStream rawOut = null;
+       RecordReader reader = getRecordReader(msg, readerFactory, session);
+       Record firstRecord = getFirstRecord(msg, reader, session);
+
+       // We have a message that we cannot parse.
+       if (firstRecord == null) {
+            if (!async) {
+              consumer.acknowledge(msg);
+            } else {
+             getAckService().submit(new Callable<Object>() {
+                 @Override
+                 public Object call() throws Exception {
+                    return consumer.acknowledgeAsync(msg).get();
```

Comment:
are we only pulling a single message from pulsar each time?  We should pull a ton of them...and write out the records..  It looks liket his is a single pulsar message but then we try to extract multiple records from it.  That is a fine path but the performant path will be to have many messages turned into a large set of records that go in a single flow file and repeat.